### PR TITLE
`ReactiveKafkaConsumer` Interface and it's Vert.x Implementation

### PR DESCRIPTION
**Pull request by debasishbsws**
Progress on #2928 

## Proposed Changes

- Identifies all the Consumer functions used and needs to be implemented in specialised modules.
- Implement all of these methods in the dispatcher vertx module with vertx implementation.

It may require some additional changes while migrating the base dispatcher to use `ReactiveKafkaConsumer`

### CODE REVIEW

File: data-plane/dispatcher/src/main/java/dev/knative/eventing/kafka/broker/dispatcher/ReactiveKafkaConsumer.java [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+ */
+public interface ReactiveKafkaConsumer<K, V> {
+
+    /**
+     * Assigns the specified partitions to this consumer.
+     *
+     * @param partitions The partitions to assign.
+     * @return A future indicating the success or failure of the assignment.
+     */
+    Future<Void> assign(Collection<TopicPartition> partitions);
```
Comment (pierDipi) [inline data-plane/dispatcher/src/main/java/dev/knative/eventing/kafka/broker/dispatcher/ReactiveKafkaConsumer.java:41]: Do we need the assign method? If `assign` is used only for testing, I think we should find a different solution for testing

Comment (debasishbsws) [reply data-plane/dispatcher/src/main/java/dev/knative/eventing/kafka/broker/dispatcher/ReactiveKafkaConsumer.java:41]: Yes, It is used only for testing. okay we can implement `assign` method on the mock we have to implement for testing.

**Comment by debasishbsws** [comment]
I added `commit(...)` method that is used in https://github.com/knative-sandbox/eventing-kafka-broker/blob/5a236cb645066be1fbb1d7ccbc95bfeac2b43ccd/data-plane/dispatcher/src/main/java/dev/knative/eventing/kafka/broker/dispatcher/impl/consumer/OffsetManager.java#L163-L165
But in apache kafkaConsumer it does not have `commit(...)`, it has `commitSync()` and `commitAsync()`

Which method signature should I use for our Interface?

File: data-plane/dispatcher/src/main/java/dev/knative/eventing/kafka/broker/dispatcher/ReactiveKafkaConsumer.java [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+ */
+public interface ReactiveKafkaConsumer<K, V> {
+
+    /**
+     * Commits the specified offsets.
+     *
+     * @param offsets The offsets to commit.
+     * @return A future containing the offsets that were committed.
+     */
+    Future<Map<TopicPartition,OffsetAndMetadata>> commit(Map<TopicPartition,OffsetAndMetadata> offset);
```
Comment (pierDipi) [inline data-plane/dispatcher/src/main/java/dev/knative/eventing/kafka/broker/dispatcher/ReactiveKafkaConsumer.java:43]:
```suggestion
    Future<Map<TopicPartition, OffsetAndMetadata>> commit(Map<TopicPartition, OffsetAndMetadata> offset);
```

File: data-plane/dispatcher-vertx/src/main/java/dev/knative/eventing/kafka/broker/dispatchervertx/VertxKafkaConsumer.java [hunk 3]
```diff
@@ -15,8 +15,104 @@
...
+            Map<TopicPartition,OffsetAndMetadata> result = new HashMap<>();
+            for(Map.Entry<io.vertx.kafka.client.common.TopicPartition,io.vertx.kafka.client.consumer.OffsetAndMetadata> entry: v.entrySet()){
+                result.put(
+                    new TopicPartition(entry.getKey().getTopic(), entry.getKey().getPartition()),
+                    new OffsetAndMetadata(entry.getValue().getOffset(), entry.getValue().getMetadata())
+                );
+            }
+            return result;
+        });
+    }
```
Comment (pierDipi) [inline data-plane/dispatcher-vertx/src/main/java/dev/knative/eventing/kafka/broker/dispatchervertx/VertxKafkaConsumer.java:66]:
```suggestion
    public Future<Map<TopicPartition, OffsetAndMetadata>> commit(Map<TopicPartition, OffsetAndMetadata> offset){

        Map<io.vertx.kafka.client.common.TopicPartition, io.vertx.kafka.client.consumer.OffsetAndMetadata> vertxOffset = new HashMap<>();

        for(Map.Entry<TopicPartition, OffsetAndMetadata> entry: offset.entrySet()){
            vertxOffset.put(
                new io.vertx.kafka.client.common.TopicPartition(entry.getKey().topic(), entry.getKey().partition()),
                new io.vertx.kafka.client.consumer.OffsetAndMetadata(entry.getValue().offset(), entry.getValue().metadata())
            );
        }
        
        return consumer.commit(vertxOffset).map( v -> {
            Map<TopicPartition, OffsetAndMetadata> result = new HashMap<>();
            for(Map.Entry<io.vertx.kafka.client.common.TopicPartition,io.vertx.kafka.client.consumer.OffsetAndMetadata> entry: v.entrySet()){
                result.put(
                    new TopicPartition(entry.getKey().getTopic(), entry.getKey().getPartition()),
                    new OffsetAndMetadata(entry.getValue().getOffset(), entry.getValue().getMetadata())
                );
            }
            return result;
        });
    }
```

File: data-plane/dispatcher-vertx/src/main/java/dev/knative/eventing/kafka/broker/dispatchervertx/VertxKafkaConsumer.java [hunk 4]
```diff
@@ -15,8 +15,104 @@
...
+        return consumer.commit(vertxOffset).map( v -> {
+            Map<TopicPartition,OffsetAndMetadata> result = new HashMap<>();
+            for(Map.Entry<io.vertx.kafka.client.common.TopicPartition,io.vertx.kafka.client.consumer.OffsetAndMetadata> entry: v.entrySet()){
+                result.put(
+                    new TopicPartition(entry.getKey().getTopic(), entry.getKey().getPartition()),
+                    new OffsetAndMetadata(entry.getValue().getOffset(), entry.getValue().getMetadata())
+                );
+            }
+            return result;
+        });
```
Comment (pierDipi) [inline data-plane/dispatcher-vertx/src/main/java/dev/knative/eventing/kafka/broker/dispatchervertx/VertxKafkaConsumer.java:65]: For both `for loop`s before commit and inside `map`, can we use the Stream API?

For example:
```java
final var vertxOffset = offset.entrySet().stream().
  map(entry -> new SimpleImmutableEntry(
      new io.vertx.kafka.client.common.TopicPartition(entry.getKey().topic(), entry.getKey().partition()),
      new io.vertx.kafka.client.consumer.OffsetAndMetadata(entry.getValue().offset(), entry.getValue().metadata())
  )).
  collect(Collectors.toMap(Entry::getKey, Entry::getValue))
```

File: data-plane/dispatcher-vertx/src/main/java/dev/knative/eventing/kafka/broker/dispatchervertx/VertxKafkaConsumer.java (hunk 3 above)
Comment (pierDipi) [reply data-plane/dispatcher-vertx/src/main/java/dev/knative/eventing/kafka/broker/dispatchervertx/VertxKafkaConsumer.java:66]: If we use the stream API (see comment below), this commit can be resolved

**Comment by pierDipi** [comment]
> I added `commit(...)` method that is used in
> 
> https://github.com/knative-sandbox/eventing-kafka-broker/blob/5a236cb645066be1fbb1d7ccbc95bfeac2b43ccd/data-plane/dispatcher/src/main/java/dev/knative/eventing/kafka/broker/dispatcher/impl/consumer/OffsetManager.java#L163-L165
> 
> 
> But in apache kafkaConsumer it does not have `commit(...)`, it has `commitSync()` and `commitAsync()`
> Which method signature should I use for our Interface?

I think the one you have used is good

**Review by pierDipi — APPROVED** [review]
/lgtm
/approve

**Pull request merged**